### ISSUE TITLE
Fn analysis not dependent on evaluation order

### ISSUE THREAD

**Issue by yuhan0** [clj-kondo/clj-kondo#1681]
**version**

clj-kondo v2022.04.25

**platform**

native

**editor**

Emacs

**problem**

The following code is valid since `(f)` is invoked before its 1-arity redefinition. However clj-kondo reports a false positive arity error, and appears to only consider the last definition of the var in the file.
```clj
(defn f [])
(f)
(defn f [x] x)
```
This is especially annoying with commented out `defn`s:

```clj
(defn foo
  ([] (foo 0)) ;; new arity
  ([x] x))

(foo) ;; <- lint error!

(comment
  ;; old version of foo for reference
  (defn foo [x] x))
```

Note: prepending the latter function with `#_:clj-kondo/ignore` doesn't suppress its analysis, only any warnings that occur in it.

**repro**

```clj
(ns repro)
(defn f [])
(f)
(comment (defn f [x] x))
```
```bash
$ clj-kondo --lint repro.clj
repro.clj:3:1: error: repro/f is called with 0 args but expects 1
linting took 236ms, errors: 1, warnings: 0
```

**config**
-NA-

**expected behavior**
I can think of two solutions:
- Analyse function calls according to preceding definitions even if they get redefined afterwards (might require a big restructuring if the current logic assumes a context-free 1-to-1 mapping between vars and their definitions?)
- Have `#_:clj-kondo/ignore` skip var analysis in the following expression such that a false positive can be suppressed. Or provide another tag for this purpose.

**Comment by borkdude** [comment]
The ignore should go before the invocation, not the function definition. 

I think this is a fairly niche use case and supporting this will complicate the logic in clj-kondo. So I currently don't think it's worth supporting.

**Comment by yuhan0** [comment]
Thanks for the reply! 
I do understand that the way it works now is for the `ignore` to go before invocations to suppress linters in the expression, but not suppress the analyser from registering findings within it. Would it be complicating the logic too much to introduce a separate keyword for this purpose? Or extend the ignore keyword to the analysis phase? 

I'm not sure how niche this use case is, but I sometimes use `comment` forms in this way as a quick way to modify a function while keeping the old one around, to be able to switch back and forth between them if needed. 

Something like the following seems reasonable (I'd be willing to help with a PR):
```clj
(defn f [^double x])
 
#_:clj-kondo/skip-analysis
(comment
  (when false
    (defn f [^int x]))
  )
  
  
(f 1.0) ;; this shouldn't lint as a type error  
(f 2)   ;; and this should!  
            
```

Note above that this could lead to false negatives as well as positives, and that there might be many calls to `f` all over the place, nested in expressions where I don't want to turn off linters entirely.

**Comment by borkdude** [comment]
@yuhan0 You can configure clj-kondo to skip over comments completely or use the `:config-in-comment` configuration to alter the configuration inside comment forms.

Another option is to use:

```
#_(f 2)
```

clj-kondo ignores those forms but you can still eval them in the REPL.

I'm not convinced this is worth supporting currently. clj-kondo just assumes one version of a function to be around at a time.
